tensor2tensor/data_generators/librispeech.py

```diff
@@ -69,41 +89,233 @@ def _get_audio_data(filepath):
+              x,
+              filters, [((1, 1), (1, 1))],
+              padding="SAME",
+              strides=(2, 1),
+              first_relu=res_relu,
+              force2d=True,
+              name="res_conv0")
+
+      # Rescale from UINT8 to floats in [-1,-1]
+      signals = (tf.to_float(inputs)-127)/128.
+      #signals = tf.contrib.framework.nest.flatten(signals)
+      signals = tf.squeeze(signals, [2, 3])
+      
+      # `stfts` is a complex64 Tensor representing the Short-time Fourier Transform of
+      # each signal in `signals`. Its shape is [batch_size, ?, fft_unique_bins]
+      # where fft_unique_bins = fft_length // 2 + 1 = 513.
+      stfts = tf.contrib.signal.stft(signals, frame_length=1024, frame_step=512,
+                               fft_length=1024)
+                               
+      # An energy spectrogram is the magnitude of the complex-valued STFT.
+      # A float32 Tensor of shape [batch_size, ?, 513].
+      magnitude_spectrograms = tf.abs(stfts)
+      
+      log_offset = 1e-6
+      log_magnitude_spectrograms = tf.log(magnitude_spectrograms + log_offset)
+      
+      # Warp the linear-scale, magnitude spectrograms into the mel-scale.
+      num_spectrogram_bins = magnitude_spectrograms.shape[-1].value
+      lower_edge_hertz, upper_edge_hertz, num_mel_bins = 80.0, 7600.0, 64
+      sample_rate = 16000
+      linear_to_mel_weight_matrix = tf.contrib.signal.linear_to_mel_weight_matrix(
+        num_mel_bins, num_spectrogram_bins, sample_rate, lower_edge_hertz,
+        upper_edge_hertz)
+      mel_spectrograms = tf.tensordot(
+        magnitude_spectrograms, linear_to_mel_weight_matrix, 1)
+      # Note: Shape inference for `tf.tensordot` does not currently handle this case.
+      mel_spectrograms.set_shape(magnitude_spectrograms.shape[:-1].concatenate(
+        linear_to_mel_weight_matrix.shape[-1:]))
+      
+      # Try without the conversion to MFCCs, first.
+      '''num_mfccs = 13
+      # Keep the first `num_mfccs` MFCCs.
+      mfccs = tf.contrib.signal.mfccs_from_log_mel_spectrograms(
+        log_mel_spectrograms)[..., :num_mfccs]'''  
+              
+      x = tf.expand_dims(mel_spectrograms, 2)
+      x.set_shape([None, None, None, num_mel_bins])
+      for i in xrange(self._model_hparams.audio_compression):
+        x = xnet_resblock(x, 2**(i + 1), True, "compress_block_%d" % i)
+      return xnet_resblock(x, self._body_input_depth, False,
+                           "compress_block_final")
+                           
+                           
+@registry.register_problem()
+class Librispeech(problem.Problem):
+  """Problem spec for English word to dictionary definition."""
+
+  @property
+  def is_character_level(self):
+    return True
+
+  @property
+  def input_space_id(self):
+    return problem.SpaceID.AUDIO_SPECTRAL
+
+  @property
+  def target_space_id(self):
+    return problem.SpaceID.EN_CHR
+
+  @property
+  def num_shards(self):

```

Comment:
transformer_base_single_gpu, does this used?

Reply:
I'm not sure I understand the question, but I used transformer_base_single_gpu as the basis for the hparams because that was what was referenced in all of the examples:

https://github.com/tensorflow/tensor2tensor/blob/master/docs/new_problem.md
https://github.com/tensorflow/tensor2tensor/blob/master/docs/walkthrough.md
https://github.com/tensorflow/tensor2tensor/blob/master/README.md